eventio/event_filter.go

```diff
@@ -10,27 +11,118 @@ type EventFilter struct {
 	OrgGUIDs   []string
 }
 
-func (filter *EventFilter) Validate() error {
-	if filter.RangeStart == "" {
-		return fmt.Errorf(`a range start filter value is required`)
+func (filter *EventFilter) SplitByMonth() ([]EventFilter, error) {
+	dateFormat := "2006-01-02"
+
+	start, err := time.Parse(dateFormat, filter.RangeStart)
+	if err != nil {
+		return nil, err
 	}
-	if filter.RangeStop == "" {
-		return fmt.Errorf(`a range stop filter value is required`)
+	end, err := time.Parse(dateFormat, filter.RangeStop)
+	if err != nil {
+		return nil, err
 	}
-	return nil
+	var filters = []EventFilter{}
+
+	nextStartOfMonth := getNextStartOfMonth(start)
+	previousStartOfMonth := getPreviousStartOfMonth(end)
+
+	if end.Before(start) {
+		return []EventFilter{}, nil
+	}
+
+	if end.Before(nextStartOfMonth) {
+		return []EventFilter{*filter}, nil
+	}
+
+	if !nextStartOfMonth.Equal(start) {
+		filters = append(filters, EventFilter{
+			RangeStart: start.Format(dateFormat),
+			RangeStop:  nextStartOfMonth.Format(dateFormat),
+			OrgGUIDs:   filter.OrgGUIDs,
+		})
+	}
+
+	for t := nextStartOfMonth; t.Before(previousStartOfMonth); t = t.AddDate(0, 1, 0) {
+		filters = append(filters, EventFilter{
+			RangeStart: t.Format(dateFormat),
+			RangeStop:  t.AddDate(0, 1, 0).Format(dateFormat),
+			OrgGUIDs:   filter.OrgGUIDs,
+		})
+	}
+
+	if !previousStartOfMonth.Equal(end) {
+		filters = append(filters, EventFilter{
+			RangeStart: previousStartOfMonth.Format(dateFormat),
+			RangeStop:  end.Format(dateFormat),
+			OrgGUIDs:   filter.OrgGUIDs,
+		})
+	}
+
+	return filters, nil
 }
 
-type PricingPlanFilter struct {
-	RangeStart string
-	RangeStop  string
+func getPreviousStartOfMonth(t time.Time) time.Time {
+	return time.Date(t.Year(), t.Month(), 1, 0, 0, 0, 0, t.Location())
+}
+
+func getNextStartOfMonth(t time.Time) time.Time {
+	beginningOfCurrentMonth := time.Date(t.Year(), t.Month(), 1, 0, 0, 0, 0, t.Location())
+	if beginningOfCurrentMonth == t {
+		return t
+	} else {
+		return beginningOfCurrentMonth.AddDate(0, 1, 0)
+	}
+}
+
+func (filter *EventFilter) TruncateMonth() (EventFilter, error) {
+	start, err := time.Parse("2006-01-02", filter.RangeStart)
+	if err != nil {
+		return *filter, err
+	}
+	stop, err := time.Parse("2006-01-02", filter.RangeStop)
+	if err != nil {
+		return *filter, err
+	}
+
+	return EventFilter{
+		RangeStart: start.Format("2006-01") + "-01",
```

Comment:
I feel this should use dates and formatting rather than string appends...

```
filterStart = time.Date(start.Year(), start.Month(), 1, 0, 0, 0, 0, start.Location())
...
RangeStart : filterStart.Format("2006-01-02")
```

Reply:
done, based on the previous introduced code. I will rebase once you are OK with it.